src/util/expr.cpp

```diff
@@ -287,6 +287,65 @@ void exprt::visit_post(std::function<void(const exprt &)> visitor) const
   visit_post_template(visitor, this);
 }
 
+optionalt<exprt>
+exprt::transform_pre(std::function<optionalt<exprt>(exprt)> visitor) const
+{
+  auto visitor_result = visitor(*this);
+
+  exprt tmp;
+
+  if(visitor_result.has_value())
+    tmp = visitor_result.value();
+  else
+    tmp = *this;
+
+  bool op_changed = false;
+
+  for(auto &op : tmp.operands()) // this breaks sharing
+  {
+    auto op_result = op.transform_pre(visitor);
+    if(op_result.has_value())
+    {
+      op = std::move(op_result.value());
+      op_changed = true;
+    }
+  }
+
+  if(op_changed)
+    return std::move(tmp);
+  else
+    return visitor_result;
+}
+
+optionalt<exprt>
+exprt::transform_post(std::function<optionalt<exprt>(exprt)> visitor) const
+{
+  exprt tmp = *this;
+  bool op_changed = false;
+
+  for(auto &op : tmp.operands()) // this breaks sharing
```

Comment:
Surely breaking sharing before you know if that's necessary will negate the benefit of this

Reply:
Yes, but note it's done on a copy.

Reply:
So what? Make a copy then immediately break sharing and you've copied the input expression